### DISCUSSION TITLE
Buttons!

### DISCUSSION THREAD

**Discussion by wibjorn** [microsoft/atlas-design#132] [Component]
Let's talk about buttons. 🔘🔼🔽🔲🔳

https://design.docs.microsoft.com/components/button.html

**Comment by brianeckmann** [comment]
Starting off the conversation here...hoping to keep the integrity of buttons since they aren't really a big issue at the moment (i.e. causing problems) so hopefully we can minimize the work and focus on other needs. We'll definitely want to touch a few things up (e.g. round the buttons a tad more). 

Consider these as suggestions based on my outsider POV from seeing a hundred other design systems (and I know much of DocsUI came from Bulma).

In a DS the difference in button types is not color, but rather the size and whether it's filled or outlined. The type signifies the hierarchy of the call to action: click this first, or click this. There shouldn't be a design that has a primary, secondary, and tertiary all in a row. More accurate use cases:

- primary button with a secondary (outlined) button—not a secondary button that's a different color (e.g. gray)
- primary button with one or more ghost or icon buttons

We want to associate filled-in buttons as primary, and ideally stick to one color (blue). Or in the case of a branded page, we have a different primary color. Secondary buttons are not filled and should be outlined or ghost, there shouldn't be a filled-in color defined as secondary or tertiary. The term "secondary" button could be removed altogether.

In the case of success, info, danger, warning, those are global colors that could be applied to other elements (e.g. successful green submission text or a text input error). There could be a use case we want to use a defined color token for a button (e.g. red danger button to delete something) but these should be defined under the approved color modifiers to buttons.

Fully rounded buttons should not be a thing, we have one value for button radius. A fully-rounded button is a different element—a clickable tag.

For sizes, the default is medium so there is no need to have both.

All in all...I don't think we need to modify a whole lot, it's more of a matter of organization and how we talk about them. I want to avoid design using different words like "secondary" in a different context than development so I'm hoping we can use the IA to help align.

**Comment by Courtny** [comment]
Wanted to shake my thoughts out on this one:

I think there is really just these _(at least to start us off)_:

1. Primary = Color: [Blue] and has three modifiers (Clear, Outlined, Filled) 
2. Danger = Color: [Red] and has three modifiers (Clear, Outlined, Filled) 
3. Muted = Color: [Gray] and has three modifiers (Clear, Outlined, Filled)

The rough image I show here is NOT actual final color values, just a quick sketch.

![671ffd53-224e-4a67-a2de-93daeee3035c](https://user-images.githubusercontent.com/1640554/118873385-b3ccc700-b8b7-11eb-9bf9-0b24a3676e6a.jpg)

**Reply by wibjorn** [reply]
Liking where this conversation is going! Few things to add:

- button padding and sizes are tied to all control sizes (so modifying them will affect `<input>` and `<select>` elements too).
- a common use case, button on a hero with a background color (uses `currentColor` to pick up whatever the text color of the section is.) probably no extra work here, just mentioning
- filled, outlined, and text sound like fine places to start
- don't design focus, docs handles keyboard focus already.
- @Courtny, muted looks troublesome for accessibility, does it definitely meet the contrast ratio?
- @brianeckmann icons can currently be placed within any button.
- @brianeckmann What's the gripe with `secondary`? can you explain?
- Do we want to only allow `primary` (blue on light theme) and `danger`, and `secondary` filled buttons? What colors to we need to support?
- What about button groups? Will we continue handling them in the same fashion?

Right now, here's the rough same of our class structure. I'm excited to pare it down a bit:

- .button _bordered, black text and border_
  - .button.is-transparent _picks up font color and uses it as text color, no border_
  - .button.is-transparent.has-inverted-border _bad naming, but same as above + border_
  - .button.is-text _has hover style, no border_
  - .button.is-<primary|secondary|tertiary|danger|warning|info|success> _filled, themed and based on our semantic colors_
  - .button.is-outlined.is-<primary|secondary|tertiary|danger|warning|info|success>_outlined, otherwise same as above_
  - .button.is-inverted.is-<primary|secondary|tertiary|danger|warning|info|success>we should remove these
  - .button.is-inverted.is-outlined.is-<primary|secondary|tertiary|danger|warning|info|success>we should remove these
  - .button.is-rounded _Let's delete! Hooray!_

**Reply by brianeckmann** [reply]
Do button padding and sizes map to a specific sizing ramp? I don't want to break anything so I'm happy to compromise if my designs change too much of the padding.

For focus...can you point me to where I can find that? I am hoping to align with Fluent vNext (upcoming big update) so maybe instead of focusing too much here I can see what you have in place.

Accessibility for disabled/muted buttons doesn't need to meet a contrast ratio.

Re: icons, some of the use cases for certain variations have some alignment issues so hopefully, we can address those.

Re: secondary gripe :D it may help in thinking about the underlying design principle. A primary button on a page should be filled in, but secondary actions should not be a filled-in button. It keeps the bad design of having more than one CTA in check. There could be a use case of needing a filled-in gray button, but it would not be labeled as 'secondary' or 'tertiary'. So in the case of the CSS framework, it's okay if you want to keep what you have but in our design conversations, a 'secondary' action will require something like outlined or transparent treatment.

Button grouping will still be needed, I think we need support for a handful of scenarios (e.g. primary with outlined and primary with transparent)

For support of buttons all up, here is my initial answer to that question. Let me know if this makes sense..
- primary by default is blue, but we have support for one neutral (gray) option along with a brand color
- we also support a filled-in button option for danger and any other use cases that warrant one of our palette colors (I don't know Docs well enough to think of another use case for something like a success or warning button...)
- outlined or transparent buttons should follow the same logic: a primary blue option, one neutral option, and support for the edge cases of danger (and others I may not know of)
- My goal is to provide flexibility for the icons and sizing but not for colors. For primary, you get the default [blue], brand, or special use case color. Otherwise, you use our specified neutral color.

Does that sound like a logical approach?

**Reply by wibjorn** [reply]
Let's discuss focus in another discussion, it a large issue, one I believe we've already handled very well.

Padding is directly tied to `control` padding. It can be modified easily, and it will update the padding on all inputs, buttons, and form controls like select.

**Reply by Courtny** [reply]
Notes from Teams thread:

In general, I was trying to keep the button naming rather specific. So you have labeling of the button type by the specific action _(Primary or Danger)_ and then a modifier depending on the emphasis needed _(`_filled, _outlined, _clear`)_ or whatever. Stick to a blue and a red button for now. And mono-chromatics (gray on white, white on black) for our themed variants. High Contrast is a special context but the primary and danger should just have a HI-C color value assigned by default.

**Reply by Courtny** [reply]
**There are certain situations where a colored button variant may be needed.** I was hoping we could figure out a clever way to have our Notifications banners recognize there is a button inside, and appropriately map a color. Some sort of a descendant button detection that has auto-mapped colors.

I would not feel comfortable just opening up these modifiers or buttons styles to every designer/dev.

Here is an example of these very specific contexts _(using outlined style)_:

![Screen Shot 2021-06-01 at 1 43 42 PM](https://user-images.githubusercontent.com/1640554/120368474-1e322e00-c2e0-11eb-9f15-654b939ba257.png)

**Reply by Courtny** [reply]
So, in summary buttons have:

1. Type of button (blue, red, monochrome)
2. Style of button (filled, outlined, clear)
3. Size of button (small, medium, large)

- We can avoid the additional context sensitive colors for now; but be considering smart solutions for those in the future to support different Semantic and Brand backgrounds.
- Icon style buttons are an extension, and Brian has it figured out. I expect iconography to work seamlessly regardless of button color, style, or sizing.

**Reply by jdanyow** [reply]
I agree with this 👆 For 1, how about: 
- "default": monochrome
- "primary": blue
- "danger": red
- "inverse": automatically chooses either black or white depending on the perceived lightness of the background. We use this type of button in cards with branded backgrounds, typically the outlined and clear styles.

**Reply by wibjorn** [reply]
@jdanyow why don't we skip `inverse` for now and circle back to it? Or just keep our current `is-transparent`?

**Comment by Courtny** [comment]
More notes summarizing design direction:

- Filled / outlined / clear for each button color.
- Only a solid blue / standard is outlined / we also have semantic colored outlined buttons
- small / medium / large (14, 16, 18 pixels) 

:hover - Darken (Light theme) / Lighten (Dark theme)
:focus - Retaining the current striped focus

**Comment by janaUX** [comment]
Adding to the button discussion... 

I think 'buttons' is the place where we don't need to get creative as a platform. 

I'd recommend aligning entirely with Fluent buttons attributes (sizes, paddings, colors, states, etc), so users benefit from carrying the same mental model across Microsoft products [More details on Fluent here](https://developer.microsoft.com/en-us/fluentui#/controls/web/button)
And we avoid having discrepancies within our platform, increasing product quality and user satisfaction as a result of that standardization.

Below are some examples of button discrepancies I found around the Docs ecosystem.

---
**1. Button size variation**
![Primary button size variations](https://user-images.githubusercontent.com/75455681/121277382-578d0f80-c885-11eb-9316-cf170e82d75c.png)

---
**2. Command button style variation/affordance**
![Icon button variations](https://user-images.githubusercontent.com/75455681/121277404-61af0e00-c885-11eb-9a36-1add530a863e.png)

---
**3. Icon button style variation** 
![Command button style variations](https://user-images.githubusercontent.com/75455681/121277420-6673c200-c885-11eb-882c-df0b0e75514b.png)

Another important point this will fix - for users who have a touch screen computer, use Docs on a tablet or phone, or have dexterity issues - they don't have the appropriate touch target area that right now is below 48x48dp.

**Reply by Courtny** [reply]
Thanks for this Jana!

**Comment by Courtny** [comment]
Here is the latest design documentation for review @jdanyow @wibjorn 

[Figma - Buttons 1.0](https://www.figma.com/proto/MCSf9XuplN2zG0sCcqJJyq/%F0%9F%9A%A7-Buttons?page-id=205%3A610&node-id=364%3A852&viewport=1342%2C2000%2C0.5&scaling=min-zoom)

**Reply by wibjorn** [reply]
👍 As of #208 the implementation for buttons within Atlas is done and can be viewed on the documentation site ⭐

https://design.docs.microsoft.com/components/button.html

**Comment by Courtny** [comment]
Buttons, as published currently in the Atlas Figma library.

![Secondary-Buttons-August-30-2021 png](https://user-images.githubusercontent.com/1640554/131410195-0d332a62-4510-413f-b1c5-6a3ff881f6db.png)
![Primary-Buttons-August-30-2021](https://user-images.githubusercontent.com/1640554/131410197-3e77b808-35e6-493b-850e-70edd9031ab3.png)

**Reply by Courtny** [reply]
Todo: Add in semantic color buttons to Figma library at later date.
